File: src/core/geo/distance.ts
import type { GeoPoint } from "@/types/geo";

const EARTH_RADIUS_KM = 6371.0088;

const toRadians = (degrees: number) => (degrees * Math.PI) / 180;

export const haversineDistanceKm = (from: GeoPoint, to: GeoPoint) => {
  const dLat = toRadians(to.latitude - from.latitude);
  const dLon = toRadians(to.longitude - from.longitude);
  const lat1 = toRadians(from.latitude);
  const lat2 = toRadians(to.latitude);

  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos(lat1) * Math.cos(lat2) * Math.sin(dLon / 2) * Math.sin(dLon / 2);

  return 2 * EARTH_RADIUS_KM * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
};

export const distanceMeters = (from: GeoPoint, to: GeoPoint) => haversineDistanceKm(from, to) * 1000;

export const roundDistanceKm = (distanceKm: number) => Math.round(distanceKm * 100) / 100;

File: lib/core/utils/geo_discovery/distance.dart
import 'dart:math' as math;

import 'geo_types.dart';

const _earthRadiusKm = 6371.0088;

double _toRadians(double degrees) => degrees * math.pi / 180;

double haversineDistanceKm(GeoPoint from, GeoPoint to) {
  final dLat = _toRadians(to.latitude - from.latitude);
  final dLon = _toRadians(to.longitude - from.longitude);
  final lat1 = _toRadians(from.latitude);
  final lat2 = _toRadians(to.latitude);

  final a =
      math.sin(dLat / 2) * math.sin(dLat / 2) +
      math.cos(lat1) * math.cos(lat2) * math.sin(dLon / 2) * math.sin(dLon / 2);

  return 2 * _earthRadiusKm * math.atan2(math.sqrt(a), math.sqrt(1 - a));
}

double distanceMeters(GeoPoint from, GeoPoint to) =>
    haversineDistanceKm(from, to) * 1000;

double roundDistanceKm(double distanceKm) =>
    (distanceKm * 100).roundToDouble() / 100;
